src/irmin-pack/inode.ml

```diff
@@ -919,7 +1074,7 @@ struct
       in
       stabilize Total t
 
-    let save layout ~add ~mem t =
+    let save layout ~add ~index t =
```

Comment:
> The saving recursion should stop as soon as it hits a keyed inode.

This now happens. (More or less: we actually require that the key exists and `mem t k = true` as otherwise the in-memory store will be unhappy.)

```ocaml
        | Tree n ->
            let save_dirty t k =
              let key =
                match Val_ref.inspect t.v_ref with
                | Key key -> if mem key then key else aux ~depth:(depth + 1) t
                | Hash hash -> (
                    match index (Lazy.force hash) with
                    | Some key -> if mem key then key else aux ~depth:(depth + 1) t
                    | None -> aux ~depth:(depth + 1) t)
              in
```

Is this enough to be able to unify `Lazy_loaded` and `Dirty`?

Reply:
As discussed offline, we can do this as part of a later refactoring pass to improve layout types.